### PR TITLE
Add support for building a RPM package.

### PR DESCRIPTION
**Pull request by Shaeto**
here is a short instruction https://github.com/Shaeto/machinekit/blob/rpmbuild/pkg/rpm/README.md

i can provide access to the test (armv7hl) repository

think this spec file can be used for other architectures (atm tested only under arm) and can be added to official Fedora collection of packages

### CODE REVIEW

**Comment by Shaeto** [comment]
unfortunately machinekit does not tag releases, so, current version of spec file will produce packages for master branch

**Comment by ArcEye** [comment]
Thanks.
I'll have a proper look at it tomorrow, been a long time since I used .rpms 

What is the equivalent Debian distro version to this (kernel version will give an idea)?

**Comment by Shaeto** [comment]
spec file does not depend on kernel version directly, i built test packages under fedora 27/kernel 4.13.16 and run on test hardware under fedora 27/kernel 4.14.3 with rt_preempt patch 5

i built and tested xenomai package also with xenomai 2.6.5, it is pretty old but i found only this version in opensuse srpm repository

**Comment by ArcEye** [comment]
> spec file does not depend on kernel version directly, i built test packages under fedora 27/kernel 4.13.16 and run on test hardware under fedora 27/kernel 4.14.3 with rt_preempt patch 5

Quite a new kernel then, about level with `sid` which has only just added a rt version of 4.14

> i built and tested xenomai package also with xenomai 2.6.5, it is pretty old but i found only this version in opensuse srpm repository

Yes Xenomai is liable to stay at v2 and eventually wither and die, unless someone feels an absolute need for it and does the work to support v3.
The rt-preempt kernels are good enough to provide a stable servo thread when using hardware step generation and much easier to maintain and use.

File: pkg/rpm/machinekit.spec [hunk 1]
```diff
@@ -0,0 +1,468 @@
+# Default values are --without-xenomai --with-rtpreempt
+
+%bcond_without rtpreempt
+%bcond_with xenomai
+
+%global commit master
+%global gittag GIT-TAG
+%global shortcommit master
+
+Summary: Marchine Kit
```
Comment (ArcEye) [inline pkg/rpm/machinekit.spec]: Typo

**Comment by ArcEye** [comment]
Q1.
Is this based upon the current machinekit repo, or a clone from a while back?

The reason I ask, it that it contains the rpm equivilent of a `postrm` file that I removed some while back.  

It was causing peculiar problems when doing updates, the old package was being removed, the new one installed and THEN the `postrm` file was being run, deleting the `/lib/linuxcnc` dir the new package had just installed.
If you uninstalled and then reinstalled all was fine, which it what you would imagine `apt dist-upgrade`
would do, but apparently not.

Q2.
You have a machinekit-devel package.  Is this just following rpm conventions or is there some other reason?

I removed the machinekit-dev package and incorporated it into the main machinekit package, quite some while back.
There was a real problem with the Makefile.inc and Makefile.modinc assuming the flavour of the building docker container (which had been set to posix) and then failing to build arm flavour components in `out of tree` builds in situ on the arm SD image.
(Should not cause you a problem if you are building natively however)

Also comp and instcomp are in the main package presumably, and would require what is in the devel package to work anyway, so for the sake of a few headers and a static lib, it can be combined.

Q3.
The main issue is probably support.
The dir can be incorporated into the repo without any issue, if we are not adding Fedora package builds, just facilitating those who might want to do so.

We only support Debian distros that we can test and package for.
How long will these scripts be current and who will update them?

The Fedora package names for instance in the deps are generally quite generic, but there will doubtless be some that change and require amendment.

The contents however are another question, looks like Fedora will always be a bit of a moving target and bound to hit problems with changed APIs etc just as we have in Debian, but probably sooner.

**Comment by ArcEye** [comment]
FYI, I am currently building a Fedora 27 Docker image to test this inside.

**Comment by Shaeto** [comment]
A1: i copied from deb repo only lines related to pru_generic.* and pru_decamux.* sym-linking to posix/rt-preempt and xenomai linuxcnc subfolders, i am not sure if these lines are necessary but definitely rpm handles link/unlink and folder remove correctly on package install, update and uninstall

A2: i just followed rpmbuild general and fedora recommended way, think it is easy to install additional -devel package (which is requires main package) to get possibility to develop new modules

A3: agree - fedora is a moving thing but they have very good release strategy and package building platform. think it is possible to contact fedora package contribution team and ask to add machinekit to official building tree and lately to official repo, i am not sure how many people will be interesting in fedora based machinekit, but i very liked their arm build :)

**Comment by ArcEye** [comment]
Congratulations, your packaging works fine.
I have just built the rpms inside a `fedora:latest` docker container, following your README
(for the most part at least)

The only issue I had was that the `dnf builddep --spec machinekit.spec` stage failed because
builddep was not installed, but fixed with `dnf install 'dnf-command(builddep)`

**Comment by ArcEye** [comment]
Thanks for your time on this, adds something new to the possibilities.

If we have any 'support issues' re the packaging, we may refer them to you 😉

**Comment by Shaeto** [comment]
no problem, i can maintain this part.

**Comment by ArcEye** [comment]
I have saved the docker I used to test, for future reference in a public repo here:
https://hub.docker.com/r/arceye/machinekit-fedora-pkg/

**Pull request merged**